core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -113,6 +140,20 @@ class KafkaMetadataLog(
     log.truncateTo(offset)
   }
 
+  override def truncateFullyToLatestSnapshot(): Boolean = {
```

Comment:
The fact that we have the high watermark in two places has always been unsatisfying, but I have not prioritized coming up with a solution. I have been inclined to say that `ReplicatedLog` owns the high watermark, so option 1 is consistent with that. I might be tempted to leave that for a separate PR though. In this case, since we are requesting a full truncation, it seems we could just set the high watermark to the snapshot offset?

Reply:
> The fact that we have the high watermark in two places has always been unsatisfying, but I have not prioritized coming up with a solution.

Yeah. I thought about it a bit when thinking about this PR. High watermark is both a property of the local log. For example, locally the high watermark cannot be less than the log start offset or greater than the log end offset.

But it is also a replicated and consensus value based on the log end offset of all the voters. For example it is technically possible for a leader to send a Fetch response to a voter or observer where the high watermark is greater than the follower's log end offset.

I agree that it would be interesting to simplify some of the code by not tracking the high watermark in two places. I'll file an issue for this.